lib/rust/ensogl/component/grid-view/src/lib.rs

```diff
@@ -587,6 +600,48 @@ where
         let column_widths = &self.widget.model().column_widths;
         entry::visible::position(row, column, self.entries_size.value(), column_widths)
     }
+
+    /// Return the size of the Entry instance for given row and column.
+    pub fn entry_size(&self, _row: Row, column: Col) -> Vector2 {
+        let column_widths = &self.widget.model().column_widths;
+        let base_entry_size = self.entries_size.value();
+        Vector2(base_entry_size.x + column_widths.width_diff(column), base_entry_size.y)
+    }
+
+    /// Return the position of the top-left corner of a viewport containing the area around the
+    /// entry at given row and col. The area around an entry is defined as the bounding box of the
+    /// entry enlarged by given margins. If there is more than one such viewport possible, return
+    /// the one closest to the current viewport.
+    fn viewport_containing_entry(&self, row: Row, column: Col, margins: Margins) -> Vector2 {
+        let pos = self.entry_position(row, column);
+        let half_size = self.entry_size(row, column) / 2.0;
+        let top = pos.y + half_size.y;
+        let bottom = pos.y - half_size.y;
+        let left = pos.x - half_size.x;
+        let right = pos.x + half_size.x;
+        let viewport = self.viewport.value();
+        let min_viewport_top = top + margins.top;
+        let max_viewport_bottom = bottom - margins.bottom;
+        let new_viewport_top = if viewport.top < min_viewport_top {
+            min_viewport_top
+        } else if viewport.bottom > max_viewport_bottom {
+            let viewport_height = viewport.size().y;
+            max_viewport_bottom + viewport_height
+        } else {
+            viewport.top
+        };
+        let min_viewport_right = right + margins.right;
+        let max_viewport_left = left - margins.left;
+        let new_viewport_left = if viewport.right < min_viewport_right {
+            let viewport_width = viewport.size().x;
+            min_viewport_right - viewport_width
+        } else if viewport.left > max_viewport_left {
+            max_viewport_left
+        } else {
+            viewport.left
+        };
+        Vector2(new_viewport_left, new_viewport_top)
```

Comment:
This function deserves detailed testing. To make it simpler, you can extract it to a separate function not taking self (only all variables needed for computation) - similarly to the `entry::visible::position` method. You can put it there: in `entry::visible` module.